addons/controllers/secretgen-controller/secretgencontrollerconfig_controller.go

```diff
@@ -0,0 +1,238 @@
+
+	// add owner reference to addon secret
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               "Cluster",
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		Controller:         pointer.BoolPtr(true),
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+	}
+
+	if !clusterapiutil.HasOwnerRef(secretGenConfig.OwnerReferences, ownerReference) {
+		log.Info("Adding owner reference to SecretGenControllerConfig")
+		secretGenConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(secretGenConfig.OwnerReferences, ownerReference)
+		*patchCRD = true
+	}
+
+	addonDataValuesSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretNameFromAddonNames(secretGenConfig.Name, constants.SecretGenControllerAddonName),
+			Namespace: secretGenConfig.Namespace,
+		},
+	}
+
+	addonDataValuesSecretMutateFn := func() error {
+		addonDataValuesSecret.Type = corev1.SecretTypeOpaque
+		addonDataValuesSecret.Data = map[string][]byte{}
+
+		// marshall the yaml contents
+		secretGenConfigYaml := &SecretGenControllerConfigSpecYaml{Namespace: secretGenConfig.Spec.Namespace, CreateNamespace: secretGenConfig.Spec.CreateNamespace}
+
+		yamlBytes, err := yaml.Marshal(secretGenConfigYaml)
+		if err != nil {
+			return err
+		}
+
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
+		addonDataValuesSecret.Data[constants.DataValueFileName] = dataValueBytes
+
+		return nil
+	}
+
+	result, err := controllerutil.CreateOrPatch(ctx, clusterClient, addonDataValuesSecret, addonDataValuesSecretMutateFn)
+	if err != nil {
+		log.Error(err, "Error creating or patching addon data values secret")
+		return err
+	}
+
+	log.Info(fmt.Sprintf("Resource %s data values secret %s", constants.SecretGenControllerAddonName, result))
+
+	return nil
+}
+
+// ReconcileSecretGenControllerConfigDelete reconciles SecretGenControllerConfig deletion
+func (r *SecretGenControllerConfigReconciler) ReconcileSecretGenControllerConfigDelete(
+	ctx context.Context,
+	log logr.Logger,
+	clusterClient client.Client,
+	secretGenConfig *addonsv1alpha1.SecretGenControllerConfig,
+	patchCRD *bool) (retErr error) {
+
+	// delete data value secret
+	addonDataValuesSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretNameFromAddonNames(secretGenConfig.Name, constants.SecretGenControllerAddonName),
+			Namespace: secretGenConfig.Namespace,
+		},
+	}
+	if err := clusterClient.Delete(ctx, addonDataValuesSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("SecretGenControllerConfig data values secret not found")
+			return nil
+		}
+		log.Error(err, "Error deleting SecretGenControllerConfig data values secret")
+		return err
+	}
+
+	// Remove finalizer from addon secret
+	*patchCRD = util.RemoveFinalizerFromCRD(log, constants.SecretGenControllerAddonName, secretGenConfig)
```

Comment:
Why do we need to patch CRD eventhough it is just getting deleted?

Reply:
If the finalizer is not removed from secretGenConfig, it won't be deleted.

Reply:
True, convinced